### PR TITLE
MIOpen FP8 and BFP8 enablement

### PR DESCRIPTION
**Pull request by JehandadKhan**
This PR enables the F8 data type in MIOpen. 

* This means addition of the `miopenFloat8` and the `miopenBFloat8` members to the `miopenDataType` enum. 
* This PR also adds the casting type the tensor descriptor and the corresponding API to get and set it. When set the kernel will cast the fp16 input tensors to the set cast type ( either miopenFloat8 or miopenBFloat8) before performing the computation.
* Currently, no performant kernels are added for the above data types or supporting casting operations. Such kernels would be added in a subsequent PR. Currently, only naive conv kernels are available and the use of rocBlas for performing F8 convolution operations.
* Only convolution operations are supported.
* All old solvers are disabled for the new data types.

### CODE REVIEW

File: src/ocl/tensorocl.cpp [hunk 1]
```diff
@@ -1446,12 +1446,12 @@ void SetTensor(const Handle& handle,
                                           std::multiplies<std::size_t>());
 
         std::size_t wld = 256 < wgd ? 256 : wgd;
-
-        std::string parms = "-DSUBTENSOR_OP_WITH_SCALAR=SUBTENSOR_OP_WITH_SCALAR_SET" +
-                            GetDataTypeKernelParams(dataType);
+        std::stringstream ss;
```
Comment (CAHEK7) [inline src/ocl/tensorocl.cpp:1475]: I'm just curious - what was the reason to replace std::string with std::stringstream? Potentially `std::stringstream` can be less efficient, more complicated, and we don't need any advanced formatting features and compatibility with stream interface since we pass `std::string` everywhere. (it's not about this particular file, it's almost everywhere)

Comment (JehandadKhan) [reply src/ocl/tensorocl.cpp:1475]: Perhaps @atamazov  can shed more light on which is efficient, but so far the insistence of reviewers has been to prefer `std::stringstream` when constructing strings since it avoid creating many intermediate copies of strings by the `+` operator. Therefore, whenever I touch older code utilizing the `+` operator to concatenate strings I prefer to replace it with stream which can "stream in" substrings without creating copies.

File: include/miopen/miopen.h [hunk 2]
```diff
@@ -690,7 +700,19 @@ MIOPEN_EXPORT miopenStatus_t miopenSetTensorDescriptor(miopenTensorDescriptor_t
                                                        const int* dimsA,
                                                        const int* stridesA);
 
-/*! @brief Get size of N-dimensional tensor
+/*! @brief Set the tensor cast type
```
Comment (JehandadKhan) [inline include/miopen/miopen.h:714]: [anchor]

**Comment by JehandadKhan** [comment]
@averinevg  Can you help update this PR to both resolve the merged conflict and update it for the upcoming `conv::ProblemDescription` changes ?

**Comment by averinevg** [comment]
> @averinevg Can you help update this PR to both resolve the merged conflict and update it for the upcoming `conv::ProblemDescription` changes ?

@JehandadKhan Sure

**Comment by averinevg** [comment]
@JehandadKhan I've resolved merge conflicts and updated this PR for the upcoming `conv::ProblemDescription` changes (https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2251/commits/53c648e1abb517046a719149c646d2d00977c32e). Please take a look at this [commit](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2251/commits/c7a079a90663a8cfe28328c8307864f591aa6025), these changes are related to https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2211. I would also suggest splitting this pull request into smaller ones.

**Comment by junliume** [comment]
@JehandadKhan  this is high urgency item while we are having some HIP tidy issues in CI.

**Comment by junliume** [comment]
@JehandadKhan some tidy issues which may need to be fixed:
```
[2023-09-05T20:16:39.003Z] /home/jenkins/workspace/MLLibs_MIOpen_jd_f8/test/gtest/conv_trans.cpp:50:12: error: enumeration values 'miopenFloat8' and 'miopenBFloat8' not explicitly handled in switch [-Werror,-Wswitch-enum]

[2023-09-05T20:16:39.003Z]     switch(prec)

[2023-09-05T20:16:39.003Z]            ^~~~

[2023-09-05T20:16:39.003Z] [ 79%] Built target test_fusion_aux

[2023-09-05T20:16:39.003Z] /home/jenkins/workspace/MLLibs_MIOpen_jd_f8/test/gtest/conv_hip_igemm_xdlops.cpp:60:12: error: enumeration values 'miopenFloat8' and 'miopenBFloat8' not explicitly handled in switch [-Werror,-Wswitch-enum]

[2023-09-05T20:16:39.003Z]     switch(prec)

[2023-09-05T20:16:39.003Z]            ^~~~
```

File: driver/conv_driver.hpp [hunk 3]
```diff
@@ -380,9 +402,14 @@ class ConvDriver : public Driver
...
         // the one of fp32 because mantissa is shorter by 13 bits.
         auto tolerance = (sizeof(Tgpu) == 4 || sizeof(Tgpu) == 1) ? 1.5e-6 : 8.2e-3;
+
         // bf16 mantissa has 7 bits, by 3 bits shorter than fp16.
         if(std::is_same<Tgpu, bfloat16>::value)
             tolerance *= 8.0;
+        constexpr bool is_fp8  = std::is_same<Tgpu, float8>::value;
+        constexpr bool is_bfp8 = std::is_same<Tgpu, bfloat8>::value;
+        if(is_bfp8 || is_fp8 || TensorsCasted())
+            tolerance = 0.3f;
```
Comment (atamazov) [inline driver/conv_driver.hpp]: It makes the `--tolerance` option useless with F8. Can we use the multiplier that, applied to the default tolerance, results in 0.3f?

File: driver/conv_driver.hpp [hunk 4]
```diff
@@ -154,6 +154,27 @@ void dumpBufferToFile(const char* fileName, T* data, size_t dataNumItems)
     }
 }
 
+miopenDataType_t DataTypeFromShortString(const std::string& type)
```
Comment (atamazov) [inline driver/conv_driver.hpp]:
```suggestion
static inline miopenDataType_t DataTypeFromShortString(const std::string& type)
```

File: driver/conv_driver.hpp [hunk 5]
```diff
@@ -154,6 +154,27 @@ void dumpBufferToFile(const char* fileName, T* data, size_t dataNumItems)
...
+
+    const auto res = conv_map.find(type);
+    if(res != conv_map.end())
+    {
+        return res->second;
+    }
+    else
+    {
+        printf("Invalid compute/cast type short hand supplied\n");
+        exit(-1);
```
Comment (atamazov) [inline driver/conv_driver.hpp]: In the driver we parse and validate options in the `ParseCmdLineArgs()` member function.

File: include/miopen/miopen.h [hunk 6]
```diff
@@ -581,6 +589,8 @@ typedef enum
     MIOPEN_CONVOLUTION_ATTRIB_DETERMINISTIC =
         1, /*!< Restrict MIOpen convolutions to kernels which produce numerically deterministic
               results. 0 - disabled (default), 1 - enabled >*/
+    MIOPEN_CONVOLUTION_ATTRIB_FP8_ROUNDING_MODE = 2,
+    MIOPEN_CONVOLUTION_ATTRIB_COMPUTE_TYPE      = 3,
```
Comment (atamazov) [inline include/miopen/miopen.h]: The documentation of these two attributes is missing. This is simple: for example, for `MIOPEN_CONVOLUTION_ATTRIB_FP8_ROUNDING_MODE` you can simply say "A value from miopenF8RoundingMode_t". But the glue should be provided.

File: driver/conv_driver.hpp [hunk 7]
```diff
@@ -1048,6 +1106,14 @@ int ConvDriver<Tgpu, Tref>::SetConvDescriptorFromCmdLineArgs()
         convDesc, spatial_dim, pads.data(), conv_strides.data(), conv_dilations.data(), mode);
 
     miopenSetConvolutionGroupCount(convDesc, group_count);
+    if(inflags.GetValueStr("compute_type") != "-1")
+    {
+        miopenDataType_t compute_type =
+            DataTypeFromShortString(inflags.GetValueStr("compute_type"));
+        // miopenSetConvolutionComputeType(convDesc, compute_type);
+        miopenSetConvolutionAttribute(
+            convDesc, MIOPEN_CONVOLUTION_ATTRIB_COMPUTE_TYPE, static_cast<int>(compute_type));
```
Comment (atamazov) [inline driver/conv_driver.hpp]: This is not implemented. I highly recommend moving this incomplete functionality to another PR.

File: include/miopen/miopen.h [hunk 8]
```diff
@@ -698,7 +708,19 @@ MIOPEN_EXPORT miopenStatus_t miopenSetTensorDescriptor(miopenTensorDescriptor_t
                                                        const int* dimsA,
                                                        const int* stridesA);
 
-/*! @brief Get size of N-dimensional tensor
+/*! @brief Set the tensor cast type
+ *
+ *  For tensors where the cast_type attribute is set, the tensor elements would be cast to the
+ * target type before the target operation is applied. Currently, only supported for convolution
+ * operations targeting the FP8 dataype
```
Comment (atamazov) [inline include/miopen/miopen.h]: typo

File: include/miopen/miopen.h [hunk 9]
```diff
@@ -698,7 +708,19 @@ MIOPEN_EXPORT miopenStatus_t miopenSetTensorDescriptor(miopenTensorDescriptor_t
                                                        const int* dimsA,
                                                        const int* stridesA);
 
-/*! @brief Get size of N-dimensional tensor
+/*! @brief Set the tensor cast type
+ *
+ *  For tensors where the cast_type attribute is set, the tensor elements would be cast to the
```
Comment (atamazov) [inline include/miopen/miopen.h]: [Recommendation] @JehandadKhan Maybe the word "conversion" is a better fit? The "cast" is something from the C/C++ programmer's and magican's vocabulary (which is not surprising because they are almost the same except wands ;)

**Review by atamazov — CHANGES_REQUESTED** [review]
Partial review

File: include/miopen/miopen.h (hunk 9 above)
Comment (JehandadKhan) [reply include/miopen/miopen.h]: Fixed!

File: include/miopen/miopen.h (hunk 8 above)
Comment (JehandadKhan) [reply include/miopen/miopen.h]: Fixed!

File: driver/conv_driver.hpp (hunk 7 above)
Comment (JehandadKhan) [reply driver/conv_driver.hpp]: Fixed!

File: include/miopen/miopen.h (hunk 6 above)
Comment (JehandadKhan) [reply include/miopen/miopen.h]: Fixed!

File: driver/conv_driver.hpp (hunk 5 above)
Comment (JehandadKhan) [reply driver/conv_driver.hpp]: Fixed!

File: driver/conv_driver.hpp (hunk 4 above)
Comment (JehandadKhan) [reply driver/conv_driver.hpp]: Fixed!

File: driver/conv_driver.hpp (hunk 3 above)
Comment (JehandadKhan) [reply driver/conv_driver.hpp]: Fixed!

File: src/kernels/MIOpenUtilKernels4.cl [hunk 10]
```diff
@@ -48,7 +48,7 @@
 #define MIOPEN_USE_INT32 0
 #endif
 
-#if MIOPEN_USE_INT8
+#if MIOPEN_USE_INT8 || MIOPEN_USE_FP8 || MIOPEN_USE_BFP8
```
Comment (junliume) [inline src/kernels/MIOpenUtilKernels4.cl:59]: @JehandadKhan `test_tensor_tran` has failed for the following issue:
```
[2023-09-09T01:57:50.415Z] MIOpen(HIP): Error [Do] 'amd_comgr_do_action(kind, handle, in.GetHandle(), out.GetHandle())' AMD_COMGR_ACTION_COMPILE_SOURCE_WITH_DEVICE_LIBS_TO_BC: ERROR (1)
[2023-09-09T01:57:50.415Z] MIOpen(HIP): Error [BuildOcl] comgr status = ERROR (1)
[2023-09-09T01:57:50.415Z] MIOpen(HIP): Warning [BuildOcl] /tmp/comgr-992752/input/MIOpenUtilKernels4.cl:51:24: error: 'MIOPEN_USE_FP8' is not defined, evaluates to 0 [-Werror,-Wundef]
[2023-09-09T01:57:50.415Z] #if MIOPEN_USE_INT8 || MIOPEN_USE_FP8 || MIOPEN_USE_BFP8
[2023-09-09T01:57:50.415Z]                        ^
[2023-09-09T01:57:50.415Z] /tmp/comgr-992752/input/MIOpenUtilKernels4.cl:51:42: error: 'MIOPEN_USE_BFP8' is not defined, evaluates to 0 [-Werror,-Wundef]
[2023-09-09T01:57:50.415Z] #if MIOPEN_USE_INT8 || MIOPEN_USE_FP8 || MIOPEN_USE_BFP8
[2023-09-09T01:57:50.415Z]                                          ^
```

File: src/kernels/hip_f8_impl.hpp [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+    }
+
+    uint32_t drop_mask = (1 << (mfmt - wm)) - 1;
+    const int max_exp  = (1 << we) - (negative_zero_nan ? 1 : 2);
+    const int exp_low_cutoff =
+        (sizeof(T) == 4 ? 128 : 16) - (1 << (we - 1)) + 1 - (negative_zero_nan ? 1 : 0);
+
+    exponent -= exp_low_cutoff - 1;
+    if(exponent <= 0)
+        drop_mask = (1 << (mfmt - wm + 1 - exponent)) - 1;
```
Comment (junliume) [inline src/kernels/hip_f8_impl.hpp]: @JehandadKhan some other tests has shown the shift may be too large in some cases: e.g.
```
runtime error: shift exponent 141 is too large for 32-bit type 'int'
```

File: src/check_numerics.cpp [hunk 12]
```diff
@@ -53,30 +53,41 @@ struct CheckNumericsResult
...
-    const std::vector<size_t> vgd = {numGlobalWorkItems, size_t{1}, size_t{1}};
-    handle.AddKernel("MIOpenCheckNumerics", "", program_name, kernel_name, vld, vgd, params)(
+    const size_t threadsPerBlock = 256;
+    const size_t numBlocks       = handle.GetMaxComputeUnits() * 6;
+    const int computeStats       = (mode & CheckNumerics::ComputeStats);
+    // TODO - some constants we should get from the device:
+    std::string program_name      = "MIOpenCheckNumerics.cpp";
+    std::string kernel_name       = GetKernelName(dDesc.GetType());
+    const std::vector<size_t> vld = {size_t{threadsPerBlock}, size_t{1}, size_t{1}};
+    const std::vector<size_t> vgd = {numBlocks, size_t{1}, size_t{1}};
```
Comment (atamazov) [inline src/check_numerics.cpp:88]: [Q] `vgd[0]` was `handle.GetMaxComputeUnits() * 6 * 256` before, but how it is `handle.GetMaxComputeUnits() * 6`. Is it correct?

File: driver/conv_driver.hpp [hunk 13]
```diff
@@ -1135,11 +1243,14 @@ int ConvDriver<Tgpu, Tref>::AllocateBuffersAndCopy()
...
-    bool is_transform       = IsInputTensorTransform();
-    bool is_int8            = data_type == miopenInt8 || data_type == miopenInt8x4;
-    size_t in_sz            = GetTensorSize(inputTensor);
-    size_t wei_sz           = GetTensorSize(weightTensor);
-    size_t out_sz           = GetTensorSize(outputTensor);
+    bool is_transform = IsInputTensorTransform();
+    bool is_int8      = data_type == miopenInt8 || data_type == miopenInt8x4;
+    // Data generated for very low precision types follows the same constraints whether its fp8,
+    // bfp8 or even if the interim tensors are being casted
+    bool is_fp8   = data_type == miopenFloat8 || data_type == miopenBFloat8 || TensorsCasted();
```
Comment (atamazov) [inline driver/conv_driver.hpp:1250]: [R][clarity] `is_f8` seems more clear because fp8 is miopenFloat8.

File: driver/conv_driver.hpp [hunk 14]
```diff
@@ -381,9 +402,14 @@ class ConvDriver : public Driver
         // Computation error of fp16 is ~2^13 (=8192) bigger than
         // the one of fp32 because mantissa is shorter by 13 bits.
         auto tolerance = (sizeof(Tgpu) == 4 || sizeof(Tgpu) == 1) ? 1.5e-6 : 8.2e-3;
+
         // bf16 mantissa has 7 bits, by 3 bits shorter than fp16.
         if(std::is_same<Tgpu, bfloat16>::value)
             tolerance *= 8.0;
+        constexpr bool is_fp8  = std::is_same<Tgpu, float8>::value;
+        constexpr bool is_bfp8 = std::is_same<Tgpu, bfloat8>::value;
```
Comment (atamazov) [inline driver/conv_driver.hpp:410]: [R][clarity] `is_bf8` -- seems better visually distinguishable from `is_fp8` than `is_bfp8`

File: driver/conv_driver.hpp [hunk 15]
```diff
@@ -381,9 +402,14 @@ class ConvDriver : public Driver
...
+        if(is_bfp8 || is_fp8 || TensorsCasted())
```
Comment (atamazov) [inline driver/conv_driver.hpp:411]: [Q] If we implement tensor casting for other datatypes than F8 then `GetDefaultTolerance` will return incorrect result.

Or I do not understand the tensor_casting feature correctly?

File: driver/conv_driver.hpp [hunk 16]
```diff
@@ -1109,6 +1191,32 @@ void RanGenSubnormBuffer(T* buf, size_t size, int percentage)
     });
 }
 
+template <>
+float8 RanGenWeights()
+{
+    const auto tmp =
+        prng::gen_0_to_B(1.0) > 0.5 ? static_cast<float>(0.0) : static_cast<float>(1.0);
```
Comment (atamazov) [inline driver/conv_driver.hpp:1198]: [Q] @CAHEK7 `static_cast<float>(prng::gen_0_to_B(1))` (i.e. the generator that should produce integer 0 or 1) should produce the same result, right?

@JehandadKhan BTW `static_cast<float>(0.0)` can be replaced with just `0.0f`

File: driver/conv_driver.hpp [hunk 17]
```diff
@@ -1394,8 +1510,11 @@ int ConvDriver<Tgpu, Tref>::AllocateBuffersAndCopy()
             db_host     = tensor<Tref>(miopen::deref(biasTensor));
             for(int i = 0; i < b_sz; i++)
             {
-                b.data[i] = static_cast<Tgpu>(i % 8) + prng::gen_canonical<Tgpu>();
-                db[i]     = static_cast<Tgpu>(i % 8) + prng::gen_canonical<Tgpu>();
+                b.data[i] =
+                    static_cast<Tgpu>(i % 8) +
+                    (is_fp8 ? prng::gen_A_to_B(Data_min, Data_max) : prng::gen_canonical<Tgpu>());
```
Comment (atamazov) [inline driver/conv_driver.hpp:1515]: [R] if Data_min/max = 0/1 for non-F8 types, then 
```suggestion
                    static_cast<Tgpu>(i % 8) + prng::gen_A_to_B(Data_min, Data_max);
```
as `gen_canonical<T>()` is equivalent to `gen_A_to_B(static_cast<T>(0), static_cast<T>(1))`

@CAHEK7 right?

File: driver/driver.hpp [hunk 18]
```diff
@@ -44,6 +42,12 @@ using float16 = half_float::half;
 #include <memory>
 #include <miopen/miopen.h>
 #include <miopen/bfloat16.hpp>
+using half         = half_float::half;
+using hip_bfloat16 = bfloat16;
+#include <miopen/hip_float8.hpp>
+using float16 = half_float::half;
+using float8  = miopen_f8::hip_f8<miopen_f8::hip_f8_type::fp8>;
+using bfloat8 = miopen_f8::hip_f8<miopen_f8::hip_f8_type::bf8>;
```
Comment (atamazov) [inline driver/driver.hpp:50]: [Q] Do we really need to put these type alias declarations in the middle of includes?

**Review by junliume — APPROVED** [review]

**Comment by junliume** [comment]
@JehandadKhan thanks about the excellent work!

File: src/gemm_v2.cpp [hunk 19]
```diff
@@ -68,14 +73,126 @@
...
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wdisabled-macro-expansion"
+#pragma clang diagnostic ignored "-Wdeprecated-declarations"
+    float alpha = gemm_desc.alpha;
+    float beta  = gemm_desc.beta;
+    auto flags  = FlagsForRocblasFp32Fp16Call(gemm_desc);
+    if(gemm_desc.conv_attributes.fp8rounding_mode.Get() == miopenF8RoundingModeStochastic)
+        flags = flags | rocblas_gemm_flags::rocblas_gemm_flags_stochastic_rounding;
+
+    rb_status = // cppcheck-suppress redundantInitialization
```
Comment (atamazov) [inline src/gemm_v2.cpp:151]: [R] In order to get rid of `// cppcheck-suppress..`, simply `return rocblas_gemm_ex3(...)` directly. Create `#else` path and `return rocblas_status::rocblas_status_internal_error` there.

File: src/gemm_v2.cpp [hunk 20]
```diff
@@ -401,35 +481,61 @@ miopenStatus_t CallGemm(const Handle& handle,
...
-                rocblas_datatype::rocblas_datatype_f16_r,
-                gemm_desc.ldc,
-                static_cast<rocblas_half*>(C) + c_offset,
-                rocblas_datatype::rocblas_datatype_f16_r,
-                gemm_desc.ldc,
-                rocblas_datatype::rocblas_datatype_f32_r,
-                rocblas_gemm_algo::rocblas_gemm_algo_standard,
-                0,
-                FlagsForRocblasFp32Fp16Call(gfx90a_alt_impl));
+            const auto is_gfx94x = miopen::StartsWith(handle.GetDeviceName(), "gfx94");
```
Comment (atamazov) [inline src/gemm_v2.cpp:484]: [R] Put this at the beginning of the function
```C++
static
const auto is_ex3_api_supported = miopen::StartsWith(handle.GetDeviceName(), "gfx94")
    && USE_ROCBLAS_GEMM_EX3;
```
and use it everywhere instead of `is_gfx94x`.

File: include/miopen/miopen.h [hunk 21]
```diff
@@ -698,7 +711,19 @@ MIOPEN_EXPORT miopenStatus_t miopenSetTensorDescriptor(miopenTensorDescriptor_t
                                                        const int* dimsA,
                                                        const int* stridesA);
 
-/*! @brief Get size of N-dimensional tensor
+/*! @brief Set the tensor cast type
+ *
+ *  For tensors where the cast_type attribute is set, the tensor elements would be converted to the
+ * target type before the target operation is applied. Currently, only supported for convolution
+ * operations targeting the FP8 datatype
```
Comment (atamazov) [inline include/miopen/miopen.h:718]: [Q] Are you sure that this is supported only for FP8 output? What about BF8?

[Q] Am I right that this is supported only for FP16 input tensors? Maybe this is worth mentioning?

[Q] Is it so that conversion (by truncation, i.e. with rounding towards zero) from FP16 (E5M10) to BF8 (E5M2) is free? Because the implementation can simply skip the low byte of FP16 elements.

**Review by atamazov — COMMENTED** [review]
Detailed review, partial. Some recommendations and questions, no mandatory comments to resolve. More tomorrow.

**Comment by atamazov** [comment]
> @JehandadKhan thanks about the excellent work!

I am second to this ))

**Comment by atamazov** [comment]
@JehandadKhan @cderb Please expect many review comments, and some are going to be about redesigning of internal data structures. 🤕

**Comment by atamazov** [comment]
@JehandadKhan @cderb I am going to create a PR instead of review, because I think this will give results must faster (at the cost of slightly worse knowledge sharing).

File: src/include/miopen/datatype.hpp [hunk 22]
```diff
@@ -159,9 +172,15 @@ inline KernelBuildParameters GetDataTypeKBP(miopenDataType_t type)
...
         {"MIOPEN_USE_RNE_BFLOAT16", use_rne_bfloat16},
+        {"MIOPEN_FP8_IEEE_EXPONENT_BIAS", MIOPEN_FP8_IEEE_EXPONENT_BIAS},
+        {"MIOPEN_FP8_CLIPPING", MIOPEN_FP8_CLIPPING},
     };
     if(use_fp64 != 0)
         kbp.Define("MIOPEN_USE_FP64", use_fp64);
+    if(use_fp8 != 0)
+        kbp.Define("MIOPEN_USE_FP8", use_fp8);
+    if(use_bfp8 != 0)
+        kbp.Define("MIOPEN_USE_FP8", use_bfp8);
```
Comment (atamazov) [inline src/include/miopen/datatype.hpp:183]: Fixed in https://github.com/ROCmSoftwarePlatform/MIOpen/pull/2452/commits/95bc7175f659c262e0fca0385df518d078ebed38

File: driver/conv_driver.hpp (hunk 16 above)
Comment (CAHEK7) [reply driver/conv_driver.hpp:1198]: @atamazov `static_cast<float>(prng::gen_canonical<int>())` is even more suitable.
But getting the whole context into account we should replace everything with:
```CPP
return static_cast<float8>(static_cast<float>(std::numeric_limits<float8>::epsilon()) * (2 * prng::gen_A_to_B(-1, 1)));
```

File: src/include/miopen/datatype.hpp [hunk 23]
```diff
@@ -159,9 +172,15 @@ inline KernelBuildParameters GetDataTypeKBP(miopenDataType_t type)
         {"MIOPEN_USE_BFP16", use_bfp16},
         {"MIOPEN_USE_INT32", use_int32},
         {"MIOPEN_USE_RNE_BFLOAT16", use_rne_bfloat16},
+        {"MIOPEN_FP8_IEEE_EXPONENT_BIAS", MIOPEN_FP8_IEEE_EXPONENT_BIAS},
+        {"MIOPEN_FP8_CLIPPING", MIOPEN_FP8_CLIPPING},
```
Comment (atamazov) [inline src/include/miopen/datatype.hpp:176]: @JehandadKhan I guess you are aware of that this change invalidates KDB for all solvers that use `SetupFloats()`, right? I would recommend producing these macros only if `MIOPEN_USE_FP8` is defined to narrow impact at least (but now it is too late and I overlooked this initially).

/cc @CAHEK7

Comment (atamazov) [reply src/include/miopen/datatype.hpp:176]: @cderb Please see previous comment.
